File: scripts/process-photos.ts
import fs from 'fs/promises';
import path from 'path';
import sharp from 'sharp';
import exifr from 'exifr';
import heicConvert from 'heic-convert';

const RAW_DIR = path.join(process.cwd(), 'photos');
const PUBLIC_PHOTOS_DIR = path.join(process.cwd(), 'public', 'photos');
const DATA_FILE = path.join(process.cwd(), 'public', 'data.json');

async function processPhotos() {
    try {
        await fs.access(RAW_DIR);
    } catch {
        console.error(`Directory ${RAW_DIR} does not exist. Please create it and add photos.`);
        return;
    }

    // Ensure output directory exists
    try {
        await fs.access(PUBLIC_PHOTOS_DIR);
    } catch {
        await fs.mkdir(PUBLIC_PHOTOS_DIR, { recursive: true });
    }

    const files = await fs.readdir(RAW_DIR);
    const photos = [];

    console.log(`Found ${files.length} files in ${RAW_DIR}`);

    for (const file of files) {
        if (file.startsWith('.')) continue; // Skip hidden files

        const filePath = path.join(RAW_DIR, file);
        const ext = path.extname(file).toLowerCase();

        if (!['.jpg', '.jpeg', '.png', '.webp', '.heic', '.dng'].includes(ext)) {
            console.log(`Skipping non-image file: ${file}`);
            continue;
        }

        console.log(`Processing ${file}...`);

        try {
            let inputBuffer: Buffer | string = filePath;
            let exifInput: Buffer | string = filePath;

            if (ext === '.heic') {
                try {
                    const inputBufferRaw = await fs.readFile(filePath);
                    // Convert to JPEG for both EXIF and Sharp
                    // heic-convert returns a JPEG buffer which should contain EXIF if we are lucky, 
                    // but heic-convert might STRIP exif. 
                    // Wait, heic-convert usually just decodes the image data.
                    // If heic-convert strips EXIF, we need to read EXIF from the original HEIC buffer.

                    // Let's try reading EXIF from the original HEIC buffer first.
                    // If exifr fails on file path, maybe it works on buffer?

                    // If exifr fails on HEIC buffer too, then we have a problem.
                    // But let's assume exifr works on buffer better than path for some reason, 
                    // OR we use the converted JPEG if exifr supports extracting from the converted one (unlikely if stripped).

                    // Actually, let's try reading EXIF from the original buffer.
                    exifInput = inputBufferRaw;

                    // And prepare buffer for sharp
                    inputBuffer = await heicConvert({
                        buffer: inputBufferRaw,
                        format: 'JPEG',
                        quality: 1
                    }) as Buffer;

                } catch (e) {
                    console.error(`Failed to convert HEIC ${file}:`, e);
                    continue;
                }
            }

            // Extract EXIF
            // We pass the original buffer for HEIC, or filepath for others
            const gps = await exifr.gps(exifInput);

            if (!gps) {
                console.warn(`No GPS data found for ${file}. Skipping map placement.`);
                continue;
            }

            const { latitude, longitude } = gps;

            // Generate filenames
            const id = path.basename(file, ext);
            const thumbName = `${id}_thumb.jpg`;
            const largeName = `${id}_large.jpg`;

            // Process images
            // Thumbnail
            await sharp(inputBuffer)
                .resize(300, 300, { fit: 'cover' })
                .jpeg({ quality: 80 })
                .toFile(path.join(PUBLIC_PHOTOS_DIR, thumbName));

            // Large version
            await sharp(inputBuffer)
                .resize(1600, 1600, { fit: 'inside', withoutEnlargement: true })
                .jpeg({ quality: 85 })
                .toFile(path.join(PUBLIC_PHOTOS_DIR, largeName));

            photos.push({
                id,
                lat: latitude,
                lng: longitude,
                thumb: `/photos/${thumbName}`,
                large: `/photos/${largeName}`,
                originalName: file
            });

            console.log(`Processed ${file}: ${latitude}, ${longitude}`);

        } catch (error) {
            console.error(`Error processing ${file}:`, error);
        }
    }

    await fs.writeFile(DATA_FILE, JSON.stringify(photos, null, 2));
    console.log(`Done! processed ${photos.length} photos. Metadata saved to ${DATA_FILE}`);
}

processPhotos();

File: scripts/process_photos.py
import io
import os
from pathlib import Path
from PIL import Image, ExifTags
from datetime import datetime
import pillow_heif
import boto3
import psycopg2
from dotenv import load_dotenv

pillow_heif.register_heif_opener()

BASE_DIR = Path(__file__).resolve().parent.parent
RAW_DIR = BASE_DIR / "photos"

load_dotenv(BASE_DIR / ".env.local")

ENDPOINT = os.environ["R2_ENDPOINT"]
BUCKET = os.environ["R2_BUCKET"]
ACCESS_KEY = os.environ["R2_ACCESS_KEY_ID"]
SECRET_KEY = os.environ["R2_SECRET_ACCESS_KEY"]
DATABASE_URL = os.environ["DATABASE_URL_UNPOOLED"]

def get_decimal_from_dms(dms, ref):
    degrees = dms[0]
    minutes = dms[1]
    seconds = dms[2]

    decimal = degrees + (minutes / 60.0) + (seconds / 3600.0)

    if ref in ['S', 'W']:
        decimal = -decimal

    return decimal

def get_gps_details(image):
    exif = image.getexif()
    if not exif:
        return None
    GPSINFO = 0x8825
    gps_info = exif.get_ifd(GPSINFO)
    if not gps_info:
        return None

    gps_latitude = gps_info.get(2)
    gps_latitude_ref = gps_info.get(1)
    gps_longitude = gps_info.get(4)
    gps_longitude_ref = gps_info.get(3)

    if gps_latitude and gps_latitude_ref and gps_longitude and gps_longitude_ref:
        lat = get_decimal_from_dms(gps_latitude, gps_latitude_ref)
        lng = get_decimal_from_dms(gps_longitude, gps_longitude_ref)
        return lat, lng

    return None



def fix_encoding(s):
    if not isinstance(s, str):
        return s
    try:
        return s.encode('latin-1').decode('utf-8')
    except:
        return s

def get_caption(image):
    exif = image.getexif()
    if not exif:
        return None

    ImageDescription = 0x010E
    caption = exif.get(ImageDescription)
    if caption:
        # Decode if bytes, though usually string in getexif()
        if isinstance(caption, bytes):
            try:
                return caption.decode('utf-8').strip()
            except:
                pass
        return fix_encoding(str(caption).strip())

    exif_ifd = exif.get_ifd(0x8769)
    if exif_ifd:
        user_comment = exif_ifd.get(0x9286)
        if user_comment:
             if isinstance(user_comment, bytes):
                # UserComment starts with 8-byte encoding ID
                # often b'ASCII\x00\x00\x00'
                try:
                    if user_comment.startswith(b'ASCII\0\0\0'):
                        return user_comment[8:].decode('utf-8').strip()
                    elif user_comment.startswith(b'UNICODE\0'):
                        return user_comment[8:].decode('utf-16').strip()
                    else:
                        # Try utf-8 blindly
                        return user_comment.decode('utf-8', errors='ignore').strip()
                except:
                    pass
             return fix_encoding(str(user_comment).strip())


    return None

def get_date_taken(image):
    exif = image.getexif()
    if not exif:
        return None

    date_str = exif.get(36867) or exif.get(36868) or exif.get(306)

    if not date_str:
         exif_ifd = exif.get_ifd(0x8769)
         if exif_ifd:
             date_str = exif_ifd.get(36867) or exif_ifd.get(36868) or exif_ifd.get(306)

    if date_str:
        # Clean up null bytes or whitespace
        date_str = str(date_str).strip().replace('\0', '')
        try:
            # Standardize to ISO format
            # EXIF dates are typically YYYY:MM:DD HH:MM:SS
            dt = datetime.strptime(date_str, '%Y:%m:%d %H:%M:%S')
            return dt.isoformat()
        except ValueError:
            pass

    return None

def process_photos():
    if not RAW_DIR.exists():
        print(f"Directory {RAW_DIR} does not exist.")
        return

    s3 = boto3.client(
        "s3",
        endpoint_url=ENDPOINT,
        aws_access_key_id=ACCESS_KEY,
        aws_secret_access_key=SECRET_KEY,
        region_name="auto",
    )

    conn = psycopg2.connect(DATABASE_URL)
    cur = conn.cursor()

    missing_gps = []
    files = [f for f in RAW_DIR.iterdir() if f.is_file() and not f.name.startswith('.')]

    print(f"Found {len(files)} files in {RAW_DIR}")

    for file_path in files:
        ext = file_path.suffix.lower()
        if ext not in ['.jpg', '.jpeg', '.png', '.webp', '.heic', '.dng']:
            continue

        try:
            with Image.open(file_path) as img:
                lat_lng = get_gps_details(img)

                if not lat_lng:
                    print(f"No GPS data found for {file_path.name}. Skipping.")
                    missing_gps.append(str(file_path))
                    continue

                lat, lng = lat_lng

                caption = get_caption(img)
                date_taken = get_date_taken(img)

                file_id = file_path.stem
                thumb_name = f"{file_id}_thumb.jpg"
                large_name = f"{file_id}_large.jpg"

                img_thumb = img.copy()
                img_thumb.thumbnail((300, 300))
                if img_thumb.mode in ('RGBA', 'P'):
                    img_thumb = img_thumb.convert('RGB')

                buf_thumb = io.BytesIO()
                img_thumb.save(buf_thumb, "JPEG", quality=80)
                s3.put_object(Bucket=BUCKET, Key=thumb_name, Body=buf_thumb.getvalue(), ContentType="image/jpeg")

                img_large = img.copy()
                img_large.thumbnail((1600, 1600))
                if img_large.mode in ('RGBA', 'P'):
                    img_large = img_large.convert('RGB')

                buf_large = io.BytesIO()
                img_large.save(buf_large, "JPEG", quality=85)
                s3.put_object(Bucket=BUCKET, Key=large_name, Body=buf_large.getvalue(), ContentType="image/jpeg")

                cur.execute(
                    """
                    INSERT INTO photos
                        (friendly_name, thumb_name, large_name, original_name, lat, lon, caption, date, status)
                    VALUES
                        (%s, %s, %s, %s, %s, %s, %s, %s, 'published')
                    ON CONFLICT (friendly_name) DO UPDATE SET
                        thumb_name    = EXCLUDED.thumb_name,
                        large_name    = EXCLUDED.large_name,
                        original_name = EXCLUDED.original_name,
                        lat           = EXCLUDED.lat,
                        lon           = EXCLUDED.lon,
                        caption       = EXCLUDED.caption,
                        date          = EXCLUDED.date,
                        status        = 'published'
                    """,
                    (
                        file_id,
                        thumb_name,
                        large_name,
                        file_path.name,
                        lat,
                        lng,
                        caption or None,
                        date_taken or None,
                    ),
                )

                print(f"Processed {file_path.name}: {lat}, {lng}")

        except Exception as e:
            print(f"Error processing {file_path.name}: {e}")

    conn.commit()
    cur.close()
    conn.close()

    print("-" * 40)
    if missing_gps:
        print(f"Missing GPS data for {len(missing_gps)} photos:")
        for name in missing_gps:
            print(name)
    print("-" * 40)
    print(f"Done! Processed {len(files) - len(missing_gps)} photos.")

if __name__ == "__main__":
    process_photos()
